### PR TITLE
[DOC] Fix links in documentation for File and IO

### PR DESCRIPTION
**Pull request by peterzhu2118**
Fixes some typos and dead links.

### CODE REVIEW

**Review by BurdetteLamar — APPROVED** [review]

File: file.c [hunk 1]
```diff
@@ -6523,7 +6523,7 @@ const char ruby_null_device[] =
  *
  *  Note that file permissions are quite different from the _mode_
  *  of a file stream (\File object).
- *  See {IO Modes}[#class-IO-label-Modes].
+ *  See {IO Modes}[IO.html#class-IO-label-Modes].
```
Comment (nobu) [inline file.c:6526]: RDoc provides link notation to sections, as well as methods.
```suggestion
 *  See IO@Modes.
```
I think this is much preferable than the generated file name and section name.

File: io.c [hunk 2]
```diff
@@ -9068,19 +9068,19 @@ rb_io_set_encoding_by_bom(VALUE io)
  *    File.new('t.txt')
  *
  *  Optional argument +mode+ (defaults to 'r') must specify a valid mode
- *  see {\IO Modes}[#class-IO-label-Modes]:
+ *  see {\IO Modes}[IO.html#class-IO-label-Modes]:
```
Comment (nobu) [inline io.c:9071]:
```suggestion
 *  see IO@Modes:
```

File: io.c [hunk 3]
```diff
@@ -9068,19 +9068,19 @@ rb_io_set_encoding_by_bom(VALUE io)
...
  *
  *    File.new('t.tmp', 'w')
  *    File.new('t.tmp', File::RDONLY)
  *
  *  Optional argument +perm+ (defaults to 0666) must specify valid permissions
- *  see {File Permissions}[#class-File-label-Permissions]:
+ *  see {File Permissions}[#class-File-label-File+Permissions]:
```
Comment (nobu) [inline io.c:9077]: Also this kind of error won't occur.
```suggestion
 *  see File@Permissions:
```

File: io.c [hunk 4]
```diff
@@ -9068,19 +9068,19 @@ rb_io_set_encoding_by_bom(VALUE io)
...
  *
  *    File.new('t.tmp', File::CREAT, 0644)
  *    File.new('t.tmp', File::CREAT, 0444)
  *
  *  Optional argument +open_opts+ must specify valid open options
- *  see {IO Open Options}[#class-IO-label-Open+Options]:
+ *  see {IO Open Options}[IO.html#class-IO-label-Open+Options]:
```
Comment (nobu) [inline io.c:9083]:
```suggestion
 *  see IO@Open+Options:
```

**Review by nobu — CHANGES_REQUESTED** [review]
Same for other URL links.

**Comment by BurdetteLamar** [comment]
Re-approved.

File: file.c [hunk 5]
```diff
@@ -6523,7 +6523,7 @@ const char ruby_null_device[] =
  *
  *  Note that file permissions are quite different from the _mode_
  *  of a file stream (\File object).
- *  See {IO Modes}[#class-IO-label-Modes].
+ *  See {IO Modes}[rdoc-ref:IO@Modes].
```
Comment (nobu) [inline file.c:6526]: Isn't `IO@Modes` just enough?

Comment (peterzhu2118) [reply file.c]: Fixed.

**Comment by nobu** [comment]
Seems no `What's Here` sections in `Enumerable` and `FileTest`. 🤷

File: array.c [hunk 6]
```diff
@@ -8080,22 +8080,22 @@ rb_ary_deconstruct(VALUE ary)
...
+ *  - {Creating an Array}[rdoc-ref:Array@Methods+for+Creating+an+Array]
+ *  - {Querying}[rdoc-ref:Array@Methods+for+Querying]
+ *  - {Comparing}[rdoc-ref:Array@Methods+for+Comparing]
+ *  - {Fetching}[rdoc-ref:Array@Methods+for+Fetching]
+ *  - {Assigning}[rdoc-ref:Array@Methods+for+Assigning]
+ *  - {Deleting}[rdoc-ref:Array@Methods+for+Deleting]
+ *  - {Combining}[rdoc-ref:Array@Methods+for+Combining]
+ *  - {Iterating}[rdoc-ref:Array@Methods+for+Iterating]
+ *  - {Converting}[rdoc-ref:Array@Methods+for+Converting]
+ *  - {And more....}[rdoc-ref:Array@Other+Methods]
```
Comment (nobu) [inline array.c:8098]: In the same namespace, the class name `Array` should be able to omit.

**Pull request merged**